contracts/neutron-interchain-txs/src/contract.rs

```diff
@@ -0,0 +1,474 @@
+            counterparty_version,
+        ),
+        _ => Ok(Response::default()),
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn migrate(_deps: DepsMut, _env: Env, _msg: MigrateMsg) -> StdResult<Response> {
+    Ok(Response::default())
+}
+
+fn sudo_open_ack(
+    deps: DepsMut,
+    _env: Env,
+    port_id: String,
+    _channel_id: String,
+    _counterparty_channel_id: String,
+    counterparty_version: String,
+) -> StdResult<Response> {
+    let parsed_version: Result<OpenAckVersion, _> =
+        serde_json_wasm::from_str(counterparty_version.as_str());
+    if let Ok(parsed_version) = parsed_version {
+        INTERCHAIN_ACCOUNTS.save(
+            deps.storage,
+            port_id,
+            &Some((
+                parsed_version.address,
+                parsed_version.controller_connection_id,
+            )),
+        )?;
+        return Ok(Response::default());
+    }
+    Err(StdError::generic_err("Can't parse counterparty_version"))
+}
+
+fn sudo_response(deps: DepsMut, request: RequestPacket, data: Binary) -> StdResult<Response> {
+    deps.api.debug(
+        format!(
+            "WASMDEBUG: sudo_response: sudo received: {:?} {:?}",
+            request, data
+        )
+        .as_str(),
+    );
+    let seq_id = request
+        .sequence
+        .ok_or_else(|| StdError::generic_err("sequence not found"))?;
+
+    let payload = read_sudo_payload(deps.storage, seq_id)?;
+    deps.api
+        .debug(format!("WASMDEBUG: sudo_response: sudo payload: {:?}", payload).as_str());
+    // handle response
+    let parsed_data = parse_response(data)?;
+
+    for item in parsed_data {
+        let item_type = item.msg_type.as_str();
+        match item_type {
+            "/cosmos.staking.v1beta1.MsgUndelegate" => {
+                let out: MsgUndelegateResponse = parse_item(&item.data)?;
+                let completion_time = out
+                    .completion_time
+                    .ok_or_else(|| StdError::generic_err("failed to get completion time"))?;
+                deps.api
+                    .debug(format!("Undelegation completion time: {:?}", completion_time).as_str());
+            }
+            _ => {
+                deps.api.debug(
+                    format!(
+                        "This type of acknowledgement is not implemented: {:?}",
+                        payload
+                    )
+                    .as_str(),
+                );
+            }
+        }
+    }
+
+    Ok(Response::default())
+}
+
+fn sudo_timeout(deps: DepsMut, _env: Env, request: RequestPacket) -> StdResult<Response> {
+    deps.api
+        .debug(format!("WASMDEBUG: sudo timeout request: {:?}", request).as_str());
+    Ok(Response::default())
+}
+
+fn sudo_error(deps: Deps, _request: RequestPacket, details: String) -> StdResult<Response> {
+    deps.api
+        .debug(format!("WASMDEBUG: sudo error: {}", details).as_str());
+    Ok(Response::default())
+}
+
+pub fn save_sudo_payload(
```

Comment:
imo it would be more readable to have the storage-related fns such as this one, the other sudo one and the ones working with reply payloads in the storage.rs file

Reply:
done